src/Tuner.cpp

```diff
@@ -0,0 +1,424 @@
+*/
+
+#include "config.h"
+
+#ifdef USE_OPENCL
+#include <array>
+#include <cassert>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <map>
+#include <random>
+#include <cmath>
+#include <fstream>
+
+#include "OpenCL.h"
+#include "Tuner.h"
+#include "Utils.h"
+
+#ifdef __APPLE__
+#include <Accelerate/Accelerate.h>
+#endif
+#ifdef USE_MKL
+#include <mkl.h>
+#endif
+#ifdef USE_OPENBLAS
+#include <cblas.h>
+#endif
+
+#define TUNER_FILE_LOCAL "leelaz_tuners"
+#define MAX_ERROR 1e-4
+
+using namespace Utils;
+
+static void sgemmBatched_ref(const std::vector<float>& a, const std::vector<float>& b,
+        std::vector<float>& c, const int m, const int n, const int k,
+        const int batch_size) {
+
+    for (auto batch = 0; batch < batch_size; batch++) {
+
+        auto offset_u = batch*m*k;
+        auto offset_v = batch*n*k;
+        auto offset_m = batch*m*n;
+
+        cblas_sgemm(CblasRowMajor, CblasTrans, CblasNoTrans,
+                    m, n, k,
+                    1.0f,
+                    &a[offset_u], m,
+                    &b[offset_v], n,
+                    0.0f,
+                    &c[offset_m], n);
+    }
+}
+
+
+bool IsMultiple(const size_t a, const size_t b) {
+    return ((a/b)*b == a);
+};
+
+bool Tuner::valid_config_sgemm(Parameters p) {
+
+    if (!IsMultiple(p["MWG"], p["MDIMC"]*p["VWM"])) {
+        return false;
+    }
+    if (!IsMultiple(p["NWG"], p["NDIMC"]*p["VWN"])) {
+        return false;
+    }
+    if (!IsMultiple(p["MWG"], p["MDIMA"]*p["VWM"])) {
+        return false;
+    }
+    if (!IsMultiple(p["NWG"], p["NDIMB"]*p["VWN"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["MDIMA"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["NDIMB"])) {
+        return false;
+    }
+    if (p["MDIMC"] != p["MDIMA"]) {
+        return false;
+    }
+    if (p["NDIMC"] != p["NDIMB"]) {
+        return false;
+    }
+    if (p["SA"] != p["SB"]) {
+        return false;
+    }
+    return true;
+}
+
+
+Parameters Tuner::get_parameters_by_int(const std::vector<Configurations> opts, const int n) {
+
+        Parameters param;
+        std::vector<size_t> choices(opts.size());
+
+        auto cfgs = 1;
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            choices[c] = opts[c].second.size();
+            cfgs *= choices[c];
+        }
+        int j = n;
+
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            auto o = opts[c];
+            auto s = o.first;
+            auto v = o.second[j % choices[c]];
+            j /= choices[c];
+            param[s] = v;
+        }
+
+        return param;
+}
+
+std::string Tuner::parameters_to_string(const Parameters p) {
+    std::string s;
+    for (auto const& x : p) {
+        s += " -D" + x.first + "=" + std::to_string(x.second);
+    }
+    return s;
+}
+
+size_t next_power_of_two(const size_t x) {
```

Comment:
This is faster.
```
size_t r = 1;
while (x >>= 1) {
    r <<= 1;
}
return  r << 1;
```

Reply:
These two implementations differ on powers of 2, though. The original leaves `x` unchanged whereas this one doubles the number. I am not sure if this matters.

Reply:
Of course x cannot be const and then is a local variable so it will be change inside this function only. and why do you say the results differ?

Reply:
[This](https://paste.fulltxt.net/3,F=uf) prints [that](https://paste.fulltxt.net/xXM@O38_7d9) on my machine.

Reply:
You are right try this: 

size_t next_power_of_two_loop(size_t x) {
    size_t r = 1;
    r <<= ( x % 2);
    while (x >>= 1) {
        r <<= 1;
    }
    return r ;
}

Reply:
And this on top fixes the equation for x = 1.

Reply:
This does not seem to work either (for even numbers that are not powers of 2)… Something like
```C
x--;
x |= x >> 1;
x |= x >> 2;
x |= x >> 4;
x |= x >> 8;
x |= x >> 16;
x |= x >> 32;
return x + 1;
```
should be okay, assuming `size_t` is 64 bits.

Reply:
Speed isn't really that important here. I'm leaving the current implementation since it works and is more readable than alternatives.

Reply:
This should work but I need to see if it is better that the original (except the erron when x is 1)
```
size_t next_power_of_two_loop(const size_t x) {
    size_t r = 1;
    size_t n = x;
    while (n >>= 1) {
        r <<= 1;
    }
    r <<= (x != r)  ;
    return r;
}
```